kpops/utils/yaml.py

```diff
@@ -63,26 +72,43 @@ def substitute_nested(input: str, **kwargs) -> str:
         }
     >>> input = "${a}, ${b}, ${c}, ${d}"
     >>> print("Substituted string: " + substitute_nested(input, **substitution))
-    0, 0, 0, 0
+    "0, 0, 0, 0"
 
     :param input: The raw input containing $-placeholders
     :param **kwargs: Substitutions
-    :raises Exception: An infinite loop condition detected. Check substitution variables.
+    :raises ValueError: An infinite loop condition detected. Check substitution variables.
     :return: Substituted input string
     """
     if not kwargs:
         return input
+    kwargs = substitute_in_self(kwargs)
     old_str, new_str = "", substitute(input, kwargs)
     steps = set()
     while new_str not in steps:
         steps.add(new_str)
         old_str, new_str = new_str, substitute(new_str, kwargs)
-    if new_str != old_str:
-        msg = "An infinite loop condition detected. Check substitution variables."
-        raise ValueError(msg)
+    _diff_substituted_str(new_str, old_str)
     return old_str
 
 
+def substitute_in_self(input: dict[str, Any]) -> dict[str, Any]:
+    """Substitute all self-references in mapping.
+
+    Will make as many passes as needed to substitute all possible placeholders.
+
+    :param input: Mapping containing $-placeholders
+    :raises ValueError: An infinite loop condition detected. Check substitution variables.
+    :return: Substituted input mapping as dict
+    """
+    old_str, new_str = "", substitute(json.dumps(input), input)
```

Comment:
could you give a little insight what's the general idea here? this looks very similar to the existing substitution function, only that we're converting to JSON first. Wouldn't it be easier to just do `model_dump_json` in `BaseDefaultsComponent`?

Reply:
It resolves the self-references in a dict. Before this PR, "chained" placeholders in the substitution mapping would get only partially substituted in the input string.

I chose to do that in a separate function, because I would have to add complexity to `substitute_nested` to differentiate between the cases of 'self-substitution' and 'normal substitution'. It seemed needlessly tangled to me. If I wanted to avoid the checks, I would have to add the code from `substitute_in_self` to `substitute_nested`. That would only lead again to worse readability and would make the function more complex without any benefits.

`substitute_in_self` is mostly useful for preparing the `kwargs` in `substitute_nested`. When substituting in a component, they are a flat dict that looks like this:

```
{
    ...
    'component.name': 'name',
    'component.type': 'type',
    'component.app.section.leaf': 'value',
    ...
}
```